### ISSUE TITLE
Training Color Adapter does not learn

### ISSUE THREAD

**Issue by paudom** [TencentARC/T2I-Adapter#97]
Hi,

I would like some help and ideas surrounding training an Adapter conditioned on spatial palette.
So far I have the following code, using current diffusers code. Could anyone give some insights on why the training, the loss is so noisy and it does not seem to train?

```python
import gc
import logging
import math
import wandb
import hydra
import os
import random
import torch
import torch.nn.functional as F

from pathlib import Path
from hydra.utils import call, instantiate
from omegaconf import OmegaConf
from accelerate import Accelerator
from tqdm.auto import tqdm
from transformers import CLIPTextModel, AutoTokenizer
from diffusers import (
    AutoencoderKL,
    DDPMScheduler,
    StableDiffusionAdapterPipeline,
    T2IAdapter,
    UNet2DConditionModel,
)
from diffusers.optimization import get_scheduler

from data.transforms import ToPILImage

def perform_validation(cfg, vae, unet, t2iadapter, accelerator, val_dataloader, *args):
    convert_to_pil = ToPILImage(cfg.dataset.transforms.mean, cfg.dataset.transforms.std)
    pipeline = StableDiffusionAdapterPipeline.from_pretrained(
        cfg.model.base_diffusion_ckpt_path,
        vae=vae,
        unet=unet,
        adapter=accelerator.unwrap_model(t2iadapter),
    ).to(accelerator.device)
    pipeline.set_progress_bar_config(disable=True)
    generator = torch.Generator(device=accelerator.device).manual_seed(cfg.seed)
    image_logs = []
    with torch.no_grad():
        for val_batch in val_dataloader:
            generated_image = pipeline(
                prompt=val_batch['caption'],
                image=val_batch['condition'],
                num_inference_steps=30,
                guidance_scale=7.5,
                generator=generator,
            ).images[0]
            image_logs.append(generated_image)
    if hasattr(accelerator, "trackers"):
        for tracker in accelerator.trackers:
            if tracker.name == "wandb":
                tracker.log({f"Validation at {val_type} {global_step}": [
                    wandb.Image(
                        image_to_log, caption="Generated"
                    ) for image_to_log in image_logs
                ]})
            else:
                logging.warn(f"image logging not implemented for {tracker.name}")
    del pipeline
    gc.collect()
    torch.cuda.empty_cache()

def encode_prompt(prompt_batch, text_encoder, tokenizer, proportion_empty_prompts):
    captions = ["" if random.random() < proportion_empty_prompts else caption for caption in prompt_batch]
    with torch.no_grad():
        text_inputs = tokenizer(
            captions,
            padding="max_length",
            max_length=tokenizer.model_max_length,
            truncation=True,
            return_tensors="pt",
        ).input_ids
        prompt_embeds = text_encoder(
            text_inputs.to(text_encoder.device),
            output_hidden_states=True,
        )[0].to(dtype=torch.float32)
    return prompt_embeds

def collate_fn(batch):
  return {
      'original': torch.stack([x['original'] for x in batch]), # Original Image
      'condition': torch.stack([x['condition'] for x in batch]), # Condition Image for adapter
      'caption': [x['caption'] for x in batch]
}

@hydra.main(config_path="config", config_name="config", version_base='1.3')
def main(cfg):
    # Disabling tokenizer logging
    os.environ['TOKENIZERS_PARALLELISM'] = 'true' if cfg.num_workers > 0 else 'false'

    # Initialize Accelerator
    accelerator = Accelerator(
        gradient_accumulation_steps=cfg.gradient_accumulation_steps,
        log_with=cfg.logger.name
    )
    accelerator.init_trackers(
        project_name=cfg.logger.project_name, 
        config=OmegaConf.to_container(cfg.logger.config_to_push, resolve=True),
        init_kwargs={cfg.logger.name: cfg.logger.init_arguments}
    )
    logging.info('Accelerator Initialized!')
    global_step = 0

    # Load Dataset, Transforms, and Dataloaders
    # Datasets
    tfms = instantiate(cfg.dataset.transforms)
    """ Original and Color Images from Dataset are already (512,512) so the transforms are only to normalize Original Image using
    [0.5,0.5,0.5] for mean and std"""
    train_dataset = instantiate(cfg.dataset_train.source, transforms=tfms, return_as_dict=True)
    logging.info(f'Train Dataset Size: {len(train_dataset)}')
    if cfg.perform_validation:
        val_dataset = instantiate(
            cfg.dataset_val.source, transforms=tfms, return_as_dict=True
        )
        logging.info(f'Validation Dataset Size: {len(val_dataset)}')

    # Dataloaders
    train_dataloader = torch.utils.data.DataLoader(
        train_dataset, batch_size=cfg.batch_size, shuffle=True, num_workers=cfg.num_workers,
        collate_fn=collate_fn
    )
    if cfg.perform_validation:
        val_dataloader = torch.utils.data.DataLoader(
            val_dataset, batch_size=cfg.validation_batch_size, shuffle=False, num_workers=cfg.num_workers
        )

    # Load UNet, VAE, Text Encoder, Tokenizer and Scheduler
    tokenizer = AutoTokenizer.from_pretrained(cfg.model.base_diffusion_ckpt_path, subfolder="tokenizer")
    text_encoder = CLIPTextModel.from_pretrained(cfg.model.base_diffusion_ckpt_path, subfolder='text_encoder')
    vae = AutoencoderKL.from_pretrained(cfg.model.base_diffusion_ckpt_path, subfolder='vae')
    unet = UNet2DConditionModel.from_pretrained(cfg.model.base_diffusion_ckpt_path, subfolder="unet")
    scheduler = DDPMScheduler.from_config(cfg.model.base_diffusion_ckpt_path, subfolder='scheduler')
    scheduler_total_timesteps = scheduler.config.num_train_timesteps
    logging.info(f'Loaded Models from: {cfg.model.base_diffusion_ckpt_path}')

    # Load T2IAdapter
    if cfg.model.pretrained_ckpt_path:
        logging.info(f'Loading Adapter from: {cfg.model.pretrained_ckpt_path}')
        t2iadapter = T2IAdapter.from_pretrained(cfg.model.pretrained_ckpt_path)
    else:
        t2iadapter = T2IAdapter(**cfg.model.parameters)

    # Deactivate Gradients (Freeze Models)
    vae.requires_grad_(False)
    text_encoder.requires_grad_(False)

    # Load Optimizer
    learning_rate = cfg.learning_rate if not cfg.scale_lr else (
        cfg.learning_rate * cfg.batch_size * cfg.gradient_accumulation_steps * accelerator.num_processes
    )
    optimizer = torch.optim.AdamW(
        t2iadapter.parameters(),
        lr=learning_rate,
        betas=(cfg.beta1, cfg.beta2),
        weight_decay=cfg.weight_decay,
        eps=cfg.epsilon
    )
    # Scheduler and math around the number of training steps.
    num_update_steps_per_epoch = math.ceil(len(train_dataloader) / cfg.gradient_accumulation_steps)
    num_training_steps = cfg.num_epochs * num_update_steps_per_epoch
    lr_scheduler = get_scheduler(
        cfg.lr_scheduler,
        optimizer=optimizer,
        num_warmup_steps=cfg.num_warmup_steps,
        num_training_steps=num_training_steps
    )

    # Prepare everything with our `accelerator`.
    vae.to(accelerator.device, dtype=torch.float32)
    unet.to(accelerator.device, dtype=torch.float32)
    text_encoder.to(accelerator.device, dtype=torch.float32)

    t2iadapter, optimizer, train_dataloader, lr_scheduler = accelerator.prepare(
        t2iadapter, optimizer, train_dataloader, lr_scheduler
    )

    # Training Loop
    for epoch in range(cfg.num_epochs):
        progress_bar = tqdm(total=len(train_dataloader), disable=not accelerator.is_local_main_process)
        progress_bar.set_description(f'Epoch {epoch}')
        for batch in train_dataloader:
            with accelerator.accumulate(t2iadapter):
                latents = vae.encode(batch['original']).latent_dist.sample()
                latents = latents * vae.config.scaling_factor
                current_batch_size = latents.shape[0]

                # Encode Prompt
                prompt_embeds = encode_prompt(
                    batch['caption'], text_encoder, tokenizer, cfg.model.proportion_empty_prompts
                )

                # For more details about why cubic sampling is used, refer to section 3.4 of https://arxiv.org/abs/2302.08453
                timesteps = torch.rand((current_batch_size,), device=cfg.device)
                timesteps = (1 - timesteps ** 3) * scheduler_total_timesteps
                timesteps = timesteps.long().clamp(0, scheduler_total_timesteps - 1)
    
                # Sample Noise
                noise = torch.randn_like(latents).to(device=cfg.device)

                # (Forward Diffusion) Add Noise to Latents
                noisy_latents = scheduler.add_noise(latents, noise, timesteps)

                # Adapter conditioning
                down_block_additional_residuals = t2iadapter(batch['condition'])

                # Predict the noise residual
                model_pred = unet(
                    noisy_latents,
                    timesteps,
                    encoder_hidden_states=prompt_embeds,
                    down_block_additional_residuals=down_block_additional_residuals,
                ).sample

                # Get the target for loss depending on the prediction type
                if scheduler.config.prediction_type == "epsilon":
                    target = noise
                elif scheduler.config.prediction_type == "v_prediction":
                    target = scheduler.get_velocity(latents, noise, timesteps)
                else:
                    raise ValueError(f"Unknown prediction type {scheduler.config.prediction_type}")

                # MSE loss
                loss = F.mse_loss(model_pred.float(), target.float(), reduction='mean')

                accelerator.backward(loss)
                if accelerator.sync_gradients:
                    accelerator.clip_grad_norm_(t2iadapter.parameters(), cfg.max_grad_norm)
                optimizer.step()
                lr_scheduler.step()
                optimizer.zero_grad()

                # Update Progress Bar
            progress_bar.update(1)
            logs = {'loss': loss.detach().item(), 'lr': lr_scheduler.get_last_lr()[0], 'step': global_step + 1}
            progress_bar.set_postfix(logs)
            global_step += 1

        if accelerator.is_local_main_process:
            if (epoch + 1) % cfg.val_epoch_interval == 0 and cfg.perform_validation:
                perform_validation(
                    cfg, vae, unet, t2iadapter, accelerator, val_dataloader
                )
            if (epoch + 1) % cfg.save_epoch_interval == 0 or epoch == cfg.num_epochs - 1:
                accelerator.save(t2iadapter.state_dict(), f'{output_path}/adapter_epoch_{epoch + 1}.pt')
    logging.info('Training Finished!')
    logging.info(f'Outputs can be found at: {output_path}')

if __name__ == '__main__':
    main()
```
Learning rate is set to 1e-5, with a batch size of 4 and using gradient accumulation for 4 steps. Also the optimizer is set to be the AdamW to the default values and using a constant lr scheduling. Am I doing something wrong here?
For more context:
I'm using this dataset: https://huggingface.co/datasets/ChristophSchuhmann/improved_aesthetics_6.5plus
I'm using the Adapter_light from Diffusers in order to train for Color
I'm using the following for the Unet: https://huggingface.co/CompVis/stable-diffusion-v1-4

Here I show the process of the training with some validation images:
![Screenshot 2023-10-13 at 10 57 51](https://github.com/TencentARC/T2I-Adapter/assets/37597137/fbbf4e35-47e1-47cd-a92b-2ac95f866eb8)
![Screenshot 2023-10-13 at 10 59 33](https://github.com/TencentARC/T2I-Adapter/assets/37597137/6175f69e-90e2-4a43-87ec-b42dc7bb332d)
![Screenshot 2023-10-13 at 11 00 05](https://github.com/TencentARC/T2I-Adapter/assets/37597137/02459e1d-0d84-4c83-8490-48db3942566e)

Is there any chance to release code for SD or SD_XL for training for spatial palette/color?

Thanks in advance!

**Comment by paudom** [comment]
For more context: 
I'm using this dataset: https://huggingface.co/datasets/ChristophSchuhmann/improved_aesthetics_6.5plus
I'm using the Adapter_light from Diffusers in order to train for Color
I'm using the following for the Unet: https://huggingface.co/CompVis/stable-diffusion-v1-4

**Comment by MC-E** [comment]
The loss in this state is normal (similar to noise). Color information is indeed more difficult to train. It is recommended to modify the sampling of the time step to a more extreme cubic:
timesteps = torch.randint(500, 1000, (bsz,), device=latents.device).long()

**Comment by paudom** [comment]
Thanks for the answer! I have three more questions:

1. So the fact that "is not learning" has nothing to do with learning rate? I thought that maybe my learning rate was not high enough.
2. Is there any chance a `train_color.py` script could be uploaded to the repo in order for me to compare it with mine? Or do you think that all training scripts are independent of the condition and therefore could be trained suing the code above changing only:
     - The type of Adapter
     - The dataset
     - The cubic timestep sampling

Thanks in advance!

**Comment by MC-E** [comment]
Q1: We have tried a larger learning rate, such as 1e-4, but this can cause training instability leading to crashes.
Q2: Yesh, only in the color control. We believe that low-level information like color is more difficult to be erased by noise, so the adapter is more likely to be ignored during the training process.
Q3. Unfortunately, we have not yet trained the color condition with SDXL.

**Comment by paudom** [comment]
Thanks a lot! Very helpful!
I will keep you updated if I achieve something in that sense.

**Comment by shoutOutYangJie** [comment]
@paudom  have you finish trianing. I also train this model. And I find the loss is very low than you at the begin of training.
@MC-E 

![image](https://github.com/TencentARC/T2I-Adapter/assets/46734514/1a5df816-da65-4aa9-a365-153c9ac09d2e)

**Comment by yataoz** [comment]
@paudom @shoutOutYangJie 
Did you guys make it work with the color palette conditioning? I encountered the same issue with the conditioning info totally ignored by the SD model, like what @paudom posted here.
